File: test/py/derivation.py
post("disableSummary")
g1 = smiles('O', name="g1")
g2 = smiles('C', name="g2")
r = ruleGMLString("""rule [ ruleID "r" context [ node [ id 0 label "O" ] ] ]""")

##############################################################################
d = Derivation()
assert str(d) == "{ }, { }"
assert repr(d) == str(d)
d.left = [g1]
assert d.left == [g1]
assert str(d) == "{ 'g1' }, { }"

d.rule = r
assert d.rule == r
assert str(d) == "{ 'g1' }, 'r', { }"
d.right = [g2]
assert d.right == [g2]
assert str(d) == "{ 'g1' }, 'r', { 'g2' }"

d = Derivation()
d.left = [None]
assert str(d) == "{ null }, { }"
d.right = [None]
assert str(d) == "{ null }, { null }"

##############################################################################
d = Derivations()
assert str(d) == "{ } < > { }"
assert repr(d) == str(d)
d.left = [g1]
assert d.left == [g1]
assert str(d) == "{ 'g1' } < > { }"

d.rules = [r]
assert d.rules == [r]
assert str(d) == "{ 'g1' } < 'r' > { }"
d.right = [g2]
assert d.right == [g2]
assert str(d) == "{ 'g1' } < 'r' > { 'g2' }"

d = Derivations()
d.left = [None]
assert str(d) == "{ null } < > { }"
d.rules = [None]
assert str(d) == "{ null } < null > { }"
d.right = [None]
assert str(d) == "{ null } < null > { null }"

##############################################################################
d = Derivation()
d.left = [g1]
d.rule = r
d.right = [g2]

dd = Derivations(d)
assert str(d) == "{ 'g1' }, 'r', { 'g2' }"
assert str(dd) == "{ 'g1' } < 'r' > { 'g2' }"

File: libs/libmod/src/mod/Derivation.cpp
#include "Derivation.hpp"

#include <mod/graph/Graph.hpp>
#include <mod/rule/Rule.hpp>

#include <ostream>

namespace mod {

Derivation::~Derivation() = default;

std::ostream &operator<<(std::ostream &s, const Derivation &d) {
	s << "{";
	for(const auto &g : d.left) {
		if(g) s << " '" << g->getName() << "'";
		else s << " null";
	}
	s << " }";
	if(d.r) s << ", '" << d.r->getName() << "'";
	s << ", {";
	for(const auto &g : d.right) {
		if(g) s << " '" << g->getName() << "'";
		else s << " null";
	}
	s << " }";
	return s;
}


Derivation::operator Derivations() const {
	if(r) return {left, {r}, {right}};
	else return {left, {}, right};
}

Derivations::~Derivations() = default;

std::ostream &operator<<(std::ostream &s, const Derivations &d) {
	s << "{";
	for(const auto &g : d.left) {
		if(g) s << " '" << g->getName() << "'";
		else s << " null";
	}
	s << " } <";
	for(const auto &r : d.rules) {
		if(r) s << " '" << r->getName() << "'";
		else s << " null";
	}
	s << " > {";
	for(const auto &g : d.right) {
		if(g) s << " '" << g->getName() << "'";
		else s << " null";
	}
	s << " }";
	return s;
}

} // namespace mod